app/javascript/components/editor/FileEditor.tsx

```diff
@@ -1,68 +1,230 @@
-import React, {
-  forwardRef,
-  useRef,
-  useImperativeHandle,
-  useCallback,
-} from 'react'
-import { File, Keybindings, WrapSetting } from './types'
-import { ExercismMonacoEditor } from './ExercismMonacoEditor'
+import React, { useEffect, useCallback, useMemo, useRef, useState } from 'react'
+import MonacoEditor from 'react-monaco-editor'
 import * as monacoEditor from 'monaco-editor/esm/vs/editor/editor.api'
-import { useLocalStorage } from '../../utils/use-storage'
+import { listen, MessageConnection } from 'vscode-ws-jsonrpc'
+import {
+  MonacoLanguageClient,
+  CloseAction,
+  ErrorAction,
+  MonacoServices,
+  createConnection,
+  MonacoModelIdentifier,
+} from 'monaco-languageclient'
+import normalizeUrl from 'normalize-url'
+import ReconnectingWebsocket from 'reconnecting-websocket'
+import { v4 as uuidv4 } from 'uuid'
+import { initVimMode, VimMode } from 'monaco-vim'
+import { EmacsExtension } from 'monaco-emacs'
+import { Keybindings, File, WrapSetting } from './types'
+
+type FileRef = {
+  filename: string
+  model: monacoEditor.editor.ITextModel
+  state: monacoEditor.editor.ICodeEditorViewState | null
+}
 
 export type FileEditorHandle = {
-  getFile: () => File
+  getFiles: () => File[]
 }
 
-type FileEditorProps = {
-  file: File
+const SAVE_INTERVAL = 500
+
+export function FileEditor({
+  language,
+  editorDidMount,
+  onRunTests,
+  onSubmit,
+  theme,
+  keybindings,
+  files,
+  wrap,
+}: {
+  editorDidMount: (editor: FileEditorHandle) => void
   language: string
   onRunTests: () => void
   onSubmit: () => void
   theme: string
   keybindings: Keybindings
+  files: File[]
   wrap: WrapSetting
-}
+}): JSX.Element {
+  // const languageServerUrl: string = useMemo(() => {
```

Comment:
These commented lines of code are related to the language server, right? If so, it might be good to add a TODO to indicate _why_ these lines are commented.

Reply:
In a PR that I will open soon, I've extracted this to a separate hook in order for us to remove the commented out code here.